java/arcs/core/policy/PolicyTranslation.kt

```diff
@@ -1,26 +1,102 @@
+        // Each handle connection needs its own check statement.
+        node.particle.spec.connections.values
+            .filter { it.direction.canRead }
+            .map { connectionSpec ->
+                // TODO(b/157605232): This only adds checks for each input handle, but not nested
+                // fields inside the handle type.
+                Check.Assert(AccessPath(node.particle, connectionSpec), checkPredicate)
+            }
+    }
+
+    // TODO(b/157605232): Add additional claims.
+    val additionalClaims = emptyMap<RecipeGraph.Node.Particle, List<Claim>>()
+
+    return graph.copyWith(additionalClaims, additionalChecks)
 }
 
-private fun checkEgressParticles(policy: Policy, graph: RecipeGraph) {
-    val invalidEgressParticles = graph.particleNodes
-        .map { it.particle.spec  }
-        .filter { !it.isolated && it.name != policy.egressParticleName }
+/**
+ * Verifies that the given egress particle nodes match the policy. The only egress particle allowed
+ * to be used with a policy named `Foo` is an egress particle named `Egress_Foo`.
+ */
+private fun checkEgressParticles(
+    policy: Policy,
+    egressParticleNodes: List<RecipeGraph.Node.Particle>
+) {
+    val invalidEgressParticles = egressParticleNodes
+        .map { it.particle.spec }
+        .filter { it.name != policy.egressParticleName }
     if (invalidEgressParticles.isNotEmpty()) {
         throw PolicyViolation.InvalidEgressParticle(policy, invalidEgressParticles.map { it.name })
     }
 }
 
+/**
+ * Constructs the [Predicate] that is to be used for checks on egress particles in the given
+ * [Policy].
+ *
+ * For policies which don't require any redaction labels, the check is a simple one of the form
+ * `check x is allowedForEgress`.
+ *
+ * For policies with redaction labels (e.g. `label1`, `label2`, `label3`), the check looks so:
+ * ```
+ * check x is allowedForEgress
+ *   or (is allowedForEgress_label1 and is label1)
+ *   or (is allowedForEgress_label2 and is label2)
+ *   or (is allowedForEgress_label3 and is label3)
+ * ```
+ */
+private fun createCheckPredicate(policy: Policy): Predicate {
+    val allowedForEgress = labelPredicate("allowedForEgress")
+    if (policy.allRedactionLabels.isEmpty()) {
+        return allowedForEgress
+    }
+    // List of predicates of the form: allowedForEgress_X AND X.
+    val labelPredicates = policy.allRedactionLabels.sorted().map { label ->
+        labelPredicate("allowedForEgress_$label") and labelPredicate(label)
+    }
+    // OR the predicates for each redaction label together.
+    return Predicate.or(listOf(allowedForEgress) + labelPredicates)
+}
+
+/** Returns a copy of the [RecipeGraph] with extra checks and claims added to particle nodes. */
+private fun RecipeGraph.copyWith(
+    additionalClaims: Map<RecipeGraph.Node.Particle, List<Claim>>,
+    additionalChecks: Map<RecipeGraph.Node.Particle, List<Check>>
+): RecipeGraph {
+    return RecipeGraph(
```

Comment:
It looks like the edges are not being updated. I think the successors and predecessors will still refer to the nodes in the old RecipeGraph? (We should make the constructor of `RecipeGraph` private to avoid such errors. ) 

We don't need the graph nodes for the translation. Why don't you simply revert to the earlier constructor that takes a `recipe` and add some additional arguments related to policy? 

e.g. create a `PolicyDFAInformation` data class (name to be bikeshedded):

```
data class PolicyDFAInformation(
        checks: Map<Particle, List<Check> = emptyList(),
        claims: Map<Handle, LIst<Claim> = emptyList()
)
```

Use it in the constructor of `RecipeGraph`:

```
class RecipeGraph(
   recipe: Recipe,  
   policyInfo: PolicyDFAInformation
) 
```
    
Then you can just build the recipe graph with the appropriate checks and claims without having to make copy?

Also, we will need  to change the `applyPolicy` to take a recipe and policy and give a `RecipeGraph`:

```
  fun applyPolicy(policy: Policy, recipe: Recipe): RecipeGraph {
     val policyInfo = // compute using the particles and handles in recipe.
  }
```

We could also change the API a bit more if we want to emphasize that we don't need the recipe for policy translation (thereby, addressing some of the lingering concerns about trusting recipes):

```
fun Policy.getDFAInformation(particles: List<Particle>, handle: List<Handle>): PolicyDFAInformation 

fun verifyPolicy(recipe: Recipe, policy: Policy) {
  val policyInfo = policy.getDFAInformation(recipe.particles, recipe.handles)
  val graph = RecipeGraph(recipe, policyInfo)
  graph.verifyChecks()
}

```
If we agree that we can process all policies at once, we will just need to change `verifyPolicy` as follows with a precomputed `policyInfo`:

```
fun verifyPolicy(recipe: Recipe, policyInfo: PolicyDFAInfo) {
  val graph = RecipeGraph(recipe, policyInfo)
  graph.verifyChecks()
}
```
(Not super happy with the names. Change as you think is appropriate if you agree with this approach.)

Reply:
I think this idea has legs. I'm leaning towards something like this:

```
class PolicyConstraints {
  val checks: Map<ParticleSpec, List<Check>>
  val claims: Map<ParticleSpec, List<Claim>>

  static fun translatePolicy(policy: Policy, particles: List<ParticleSpec>) {...}
}
```

There's a bit of tidying up to do with RecipeGraph first though, so I'd like to add a TODO and submit this PR, and then update all the APIs in a follow-up